### ISSUE TITLE
Engine.TensorMatMul returns zeros for 2D matrix multiplication; Tensor indexer writes don't persist for Engine-created tensors

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#41]
## Summary

Two critical bugs discovered during comprehensive invariant testing of all 170+ layer types in AiDotNet. Both issues are in the AiDotNet.Tensors NuGet package and affect backward pass gradient computation and parameter serialization across many layers.

### Description
`Engine.TensorMatMul(A, B)` returns an all-zero tensor for certain 2D matrix shapes, particularly column-vector multiplications used in backward pass gradient computation.

### Reproduction
```csharp
// ReservoirLayer Forward - this produces zeros:
var inputColumn = scaledInput.Reshape([inputSize, 1]);  // [100, 1]
var result = Engine.TensorMatMul(_inputWeights, inputColumn);  // [100, 100] @ [100, 1] → all zeros!

// Workaround that produces correct results:
var result = new Tensor<T>([reservoirSize]);
for (int r = 0; r < reservoirSize; r++)
{
    T sum = NumOps.Zero;
    for (int c = 0; c < inputSize; c++)
        sum = NumOps.Add(sum, NumOps.Multiply(_inputWeights[r, c], scaledInput[c]));
    result[r] = sum;
}
```

### Affected shapes (confirmed)
- `[N, M] @ [M, 1]` → column vector multiplication (most common case)
- `[1, M] @ [M, N]` → row vector multiplication
- `[M, 1] @ [1, N]` → outer product

### Impact
- **5 layers** have zero backward gradients: GRULayer, LSTMLayer, BatchNormalizationLayer, RBMLayer, SpikingLayer
- **ReservoirLayer** forward pass returned all zeros (fixed with manual matmul workaround)
- **DeepBoltzmannMachine** PropagateUp/Down returned zeros (fixed with manual matmul)
- Any layer using `Engine.TensorMatMul` in its backward pass for gradient computation produces zero gradients, breaking training

### Note
`Engine.FusedLinear` works correctly for the same shapes (it's used in DenseLayer.Forward). The issue is specifically with `Engine.TensorMatMul`.

---

### Description
Tensors returned by Engine operations (e.g., `Engine.TensorSubtract`, `Engine.TensorMultiplyScalar`) appear to be immutable views. Writing to them via the `[i, j]` indexer, `SetFlat()`, or `Data.Span` has no effect — subsequent reads return the original values.

### Reproduction
```csharp
// Create tensor via Engine operations
var rawWeights = Tensor<T>.CreateRandom(4, 8);
var scaled = Engine.TensorMultiplyScalar(rawWeights, scaleFactor);
var weights = Engine.TensorSubtract(scaled, offsetTensor);

// These writes are silently ignored:
weights[0, 0] = NumOps.FromDouble(99.0);
weights.SetFlat(0, NumOps.FromDouble(99.0));
weights.Data.Span[0] = NumOps.FromDouble(99.0);

// Reading back returns the ORIGINAL value, not 99.0
Console.WriteLine(weights[0, 0]);  // Still the original value!
```

### Workaround
Create a fresh `new Tensor<T>(shape)` and copy values element-by-element:
```csharp
var mutableWeights = new Tensor<T>([inputSize, outputSize]);
for (int i = 0; i < inputSize; i++)
    for (int j = 0; j < outputSize; j++)
        mutableWeights[i, j] = engineTensor[i, j];  // Read works, write works on new tensor
```

### Impact
- **SpikingLayer** `SetParameters` doesn't update weights (Parameters_SetGet_Roundtrip test fails)
- Any layer that initializes weights via Engine operations and later tries to modify them in-place (e.g., during deserialization or parameter updates) silently fails
- **DenseLayer** worked around this by using `Data.Span` writes on tensors created via `new Tensor<T>(shape)` (not via Engine)

---

## Environment
- AiDotNet.Tensors version: latest on NuGet
- .NET 10.0 and .NET Framework 4.7.1
- CPU engine (no GPU)
- Discovered via LayerTestBase invariant tests (12 mathematical invariants per layer)

## Test Infrastructure
These bugs were found using a new `LayerTestBase` class that tests mathematical invariants for every layer implementation:
- Forward: finite output, deterministic, input-sensitive, shape-consistent
- Backward: finite gradients, non-zero weight gradients, numerical gradient check
- Parameters: count matches vector, set/get roundtrip
- Serialization: roundtrip preserves behavior

Current results: 123/132 layer tests passing (93%), with all 9 failures traced to these two package bugs.
